File: os/DeadlockDetection.java
import java.util.Random;
import java.util.Scanner;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class DeadlockDetection {

    private static final Lock lock1 = new ReentrantLock();
    private static final Lock lock2 = new ReentrantLock();
    private static int maxRetries = 10; // Default retry limit

    // Function to simulate random sleep time
    private static void sleepRandomTime() {
        try {
            Thread.sleep(new Random().nextInt(100)); // Sleep for 0-100 milliseconds
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    // Task function to try to acquire both locks
    private static void task(Lock firstLock, Lock secondLock, String taskName) {
        int retries = 0;
        while (retries < maxRetries) {
            if (firstLock.tryLock()) {
                System.out.println(taskName + " acquired " + (firstLock == lock1 ? "Lock1" : "Lock2") + ".");
                if (secondLock.tryLock()) {
                    System.out.println(taskName + " acquired both locks: Executing " + taskName + "...");
                    secondLock.unlock();
                    firstLock.unlock();
                    return;
                } else {
                    firstLock.unlock();
                    System.out.println(taskName + " released " + (firstLock == lock1 ? "Lock1" : "Lock2") + " as " + (secondLock == lock1 ? "Lock1" : "Lock2") + " was unavailable.");
                }
            }
            System.out.println(taskName + " retrying... Attempt " + (++retries) + "/" + maxRetries);
            sleepRandomTime();
        }
        System.out.println(taskName + " failed to acquire locks after " + maxRetries + " attempts. Exiting.");
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the maximum number of retry attempts for each task (default is 10): ");
        maxRetries = scanner.nextInt();

        Thread t1 = new Thread(() -> task(lock1, lock2, "Task 1"));
        Thread t2 = new Thread(() -> task(lock2, lock1, "Task 2"));

        t1.start();
        t2.start();

        try {
            t1.join();
            t2.join();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }

        System.out.println("\nDeadlock check:");
        System.out.println("If both tasks fail to acquire both locks after retrying " + maxRetries + " times, a deadlock is likely.");

        System.out.println("\nBoth tasks completed (or exited after reaching retry limit).");
    }
}

File: os/deadlock_detection.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h>

pthread_mutex_t lock1 = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t lock2 = PTHREAD_MUTEX_INITIALIZER;

int max_retries = 10; // Default retry limit

// Function to simulate random sleep time
void sleepRandomTime() {
    usleep((rand() % 100) * 1000); // Sleep for 0-100 milliseconds
}

// Task function to try to acquire both locks
void* task(void* arg) {
    int retries = 0;
    pthread_mutex_t *first_lock = (pthread_mutex_t *)arg;
    pthread_mutex_t *second_lock = (first_lock == &lock1) ? &lock2 : &lock1;
    char *task_name = (first_lock == &lock1) ? "Task 1" : "Task 2";

    while (retries < max_retries) {
        if (pthread_mutex_trylock(first_lock) == 0) {
            printf("%s acquired %s.\n", task_name, (first_lock == &lock1) ? "Lock1" : "Lock2");
            if (pthread_mutex_trylock(second_lock) == 0) {
                printf("%s acquired both locks: Executing %s...\n", task_name, task_name);
                pthread_mutex_unlock(second_lock);
                pthread_mutex_unlock(first_lock);
                return NULL;
            } else {
                pthread_mutex_unlock(first_lock);
                printf("%s released %s as %s was unavailable.\n", task_name, (first_lock == &lock1) ? "Lock1" : "Lock2", (second_lock == &lock1) ? "Lock1" : "Lock2");
            }
        }
        printf("%s retrying... Attempt %d/%d\n", task_name, ++retries, max_retries);
        sleepRandomTime();
    }
    printf("%s failed to acquire locks after %d attempts. Exiting.\n", task_name, max_retries);
    return NULL;
}

// Function to simulate a deadlock check
void deadlockCheck(pthread_t t1, pthread_t t2) {
    // Monitor if tasks failed to acquire locks after max retries
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);
    printf("\nDeadlock check:\n");
    printf("If both tasks fail to acquire both locks after retrying %d times, a deadlock is likely.\n", max_retries);
}

int main() {
    srand(time(NULL)); // Seed random generator

    printf("Enter the maximum number of retry attempts for each task (default is 10): ");
    scanf("%d", &max_retries);

    pthread_t t1, t2;

    // Create threads
    if (pthread_create(&t1, NULL, task, &lock1) != 0) {
        perror("Failed to create thread 1");
        return 1;
    }
    if (pthread_create(&t2, NULL, task, &lock2) != 0) {
        perror("Failed to create thread 2");
        return 1;
    }

    // Perform deadlock check after both threads finish
    deadlockCheck(t1, t2);

    printf("\nBoth tasks completed (or exited after reaching retry limit).\n");

    return 0;
}